libxrdp/xrdp_mcs.c

```diff
@@ -1006,114 +1013,86 @@ xrdp_mcs_out_gcc_data(struct xrdp_sec *self)
-                  self->crypt_method, self->crypt_level, SEC_TAG_PUBKEY,
-                  0x005c, SEC_RSA_MAGIC, 0x0048, 0x00020000, 0x3f000000,
-                  SEC_TAG_KEYSIG, 72);
-    }
-    else if (self->rsa_key_bytes == 256)
-    {
-        LOG(LOG_LEVEL_DEBUG, "using 2048 bit RSA key");
-        /* [MS-RDPBCGR] TS_UD_HEADER */
-        out_uint16_le(s, SEC_TAG_SRV_CRYPT);  /* type */
-        out_uint16_le(s, 0x01ac);             /* length is 428 */
+                  "type 0x%4.4x, length %d", SEC_TAG_SRV_CRYPT, header_length);
+
         /* [MS-RDPBCGR] TS_UD_SC_SEC1 */
-        out_uint32_le(s, self->crypt_method); /* encryptionMethod  */
-        out_uint32_le(s, self->crypt_level);  /* encryptionLevel */
-        out_uint32_le(s, 32);                 /* serverRandomLen */
-        out_uint32_le(s, 0x178);              /* serverCertLen (376 bytes) */
+        out_uint32_le(s, self->crypt_method);   /* encryptionMethod  */
+        out_uint32_le(s, self->crypt_level);    /* encryptionLevel */
+        out_uint32_le(s, 32);                   /* serverRandomLen */
+        out_uint32_le(s, server_cert_len);      /* serverCertLen */
         out_uint8a(s, self->server_random, 32); /* serverRandom */
+        LOG_DEVEL(LOG_LEVEL_TRACE, "Adding struct [MS-RDPBCGR] TS_UD_SC_SEC1 "
+                  "encryptionMethod 0x%8.8x, encryptionLevel 0x%8.8x, "
+                  "serverRandomLen 32, serverCertLen %d, serverRandom (omitted), ",
+                  self->crypt_method, self->crypt_level, server_cert_len);
+
         /* (field serverCertificate) [MS-RDPBCGR] SERVER_CERTIFICATE */
         /* HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\ */
         /* TermService\Parameters\Certificate */
         out_uint32_le(s, 1); /* dwVersion (1 = PROPRIETARYSERVERCERTIFICATE) */
+        LOG_DEVEL(LOG_LEVEL_TRACE, "Adding struct [MS-RDPBCGR] SERVER_CERTIFICATE "
+                  "dwVersion.certChainVersion 1 (CERT_CHAIN_VERSION_1), "
+                  "dwVersion.t 0 (permanent)");
+
         /* [MS-RDPBCGR] PROPRIETARYSERVERCERTIFICATE */
         out_uint32_le(s, 1);              /* dwSigAlgId (1 = RSA) */
         out_uint32_le(s, 1);              /* dwKeyAlgId (1 = RSA) */
         out_uint16_le(s, SEC_TAG_PUBKEY); /* wPublicKeyBlobType (BB_RSA_KEY_BLOB) */
-        out_uint16_le(s, 0x011c);         /* wPublicKeyBlobLen (284 bytes) */
+        out_uint16_le(s, public_key_blob_len); /* wPublicKeyBlobLen */
+
         /* (field PublicKeyBlob) [MS-RDPBCGR] RSA_PUBLIC_KEY */
         out_uint32_le(s, SEC_RSA_MAGIC);  /* magic (0x31415352 'RSA1') */
-        out_uint32_le(s, 0x0108);         /* keylen (264 bytes = (bitlen / 8) modulus + 8 padding) */
-        out_uint32_be(s, 0x00080000);     /* bitlen = 2048 */
-        out_uint32_be(s, 0xff000000);     /* datalen (255 = (bitlen / 8) - 1) */
+        out_uint32_le(s, key_len);        /* keylen */
+        out_uint32_le(s, bit_len);        /* bitlen */
```

Comment:
Thanks for changing these two calls from `out_int32_be()` - this makes a lot more sense.

Reply:
Yeah I have no idea why only those two lines had their endianness reversed, especially since they are plain length values that can't be expressed as the plain int value when using the big endian macros.

Reply:
It looks like one of those late-night coding things we've all done...